libfeed/feed-model-ordering.c

```diff
@@ -0,0 +1,456 @@
+        {
+          state->inner_outer_index = (state->inner_outer_index + 1) % INNER_OUTER_LENGTH;
+
+          /* If we wrapped around on the inner_outer_index then increase the
+           * number of cards to take from each type. This will make the number
+           * of cards on each type appear to increase as the feed scrolls down. */
+          if (state->inner_outer_index == 0)
+            ++state->n_cards_to_take;
+        }
+    }
+  else
+    {
+      /* Haven't taken too many cards for this type yet. But we might
+       * have taken the limit of cards that we can take per app for
+       * this type. If that is the case, then we'll need to move to
+       * the next app for this card type. */
+      GPtrArray *cards_source = g_ptr_array_index (card_sources_for_type,
+                                                   index_for_card_type);
+      guint card_limit_for_app = MIN (lookup_card_limit (card_type),
+                                      cards_source->len);
+
+      if (state->cards_taken_from_app >= card_limit_for_app)
+        {
+          /* We've taken as many as we can for this app, but we still
+           * have cards left to take from this type. Go to the next app
+           * index. */
+          state->cards_taken_from_app = 0;
+          g_hash_table_replace (state->app_indices_for_card_type,
+                                GINT_TO_POINTER (card_type),
+                                GINT_TO_POINTER (index_for_card_type + 1));
+        }
+    }
+}
+
+/**
+ * eos_discovery_feed_arrange_orderable_models:
+ * @unordered_orderable_models: (element-type EosDiscoveryFeedOrderableModel): The
+ *                              models to order. The output list may be truncated
+ *                              as a result of the reordering operation.
+ *
+ * Reorder the models in the order that they should be displayed in the feed.
+ * The pattern here is not predictable by the user but shuffles through a range
+ * of content adequately and is not in fact random, but changes day by day.
+ *
+ * The general approach here is to insert a news card, article card, video card
+ * and artwork card from a diferent app, with the word-quote card being inserted
+ * second. Then a news card, video card, article card and artwork card will
+ * be inserted and the pattern will repeat itself. There is a limit of one card
+ * per app, except for news cards where there is a limit of 5 cards.
+ *
+ * Returns: (transfer container) (element-type EosDiscoveryFeedOrderableModel): The
+ *          correctly ordered models.
+ */
+GPtrArray *
+eos_discovery_feed_arrange_orderable_models (GPtrArray                                   *unordered_orderable_models,
+                                             EosDiscoveryFeedArrangeOrderableModelsFlags  flags)
+{
+  g_autoptr(GHashTable) descriptor_map = arrange_descriptors_into_map (unordered_orderable_models);
+  guint overall_index = 0;
+  gboolean evergreen_card_added = FALSE;
```

Comment:
Should these two not be in the state as well?

Reply:
Not really - the evergreen card is kind of a special case which is conceptually separate to the fundamental ordering algorithm. I wanted to keep the two in separate places so that it was more clear how the algorithm worked.